File: Spark operations (e.g., transformations and actions) and MapReduce/Python/task1.py
from pyspark import SparkContext
import json
import re
import sys


def sub_tasks(review_rdd, output_file, stopwords, year, m, n):
    total_reviews = review_rdd.map(lambda x: x['review_id']).count()

    y_reviews = review_rdd.filter(lambda x: str(year) in x['date']).count()

    distinct_users = review_rdd.map(lambda x: x['user_id']).distinct().count()

    top_reviewers = review_rdd.map(lambda x: (x['user_id'], 1)) \
        .reduceByKey(lambda x, y: x + y) \
        .sortBy(lambda x: (-x[1], x[0])) \
        .take(int(m))

    pun_list = ["(", "[", ",", ".", "!", "?", ":", ";", "]", ")"]

    stopwords_list = []
    with open(str(stopwords), 'rt') as stopwords_file:
        for word in stopwords_file:
            stopwords_list.append(word.strip())

    top_words = review_rdd.flatMap(lambda x: x['text'].lower().strip().split()) \
        .filter(lambda x: x not in pun_list) \
        .map(lambda x: re.sub('[^a-zA-Z0-9\s]', '', x)) \
        .filter(lambda x: x not in stopwords_list and len(x) > 0) \
        .map(lambda x: (x, 1)) \
        .reduceByKey(lambda x, y: x + y) \
        .sortBy(lambda x: (-x[1], x[0])).map(lambda x: x[0]).take(int(n))

    output = {
        "A": total_reviews,
        "B": y_reviews,
        "C": distinct_users,
        "D": top_reviewers,
        "E": top_words,
    }

    with open(str(output_file), "w+") as fileout:
        json.dump(output, fileout, indent=4)


def main():
    if len(sys.argv) != 7:
        print("Please enter all the required agruments")
        exit(-1)

    sc = SparkContext('local[*]', 'task1')

    review_rdd = sc.textFile(str(sys.argv[1])).map(lambda x: json.loads(x)).cache()

    sub_tasks(review_rdd, sys.argv[2], sys.argv[3], sys.argv[4], sys.argv[5], sys.argv[6])

    sc.stop()


if __name__ == "__main__": main()

File: Spark operations (e.g., transformations and actions) and MapReduce/Scala/task1.scala
import java.io.{BufferedWriter, File, FileWriter}
import org.apache.spark.sql.SparkSession
import org.json4s.jackson.JsonMethods.parse
import scala.io._


object task1 {

  def main(args: Array[String]): Unit ={

    if (args.length != 6) {
      println("Please enter all the required agruments")
      sys.exit(1)
    }

    val lines = Source.fromFile(args(2)).getLines.toList
    val pun_list = List("(", "[", ",", ".", "!", "?", ":", ";", "]", ")")
    val reg_pattern = "[^a-zA-Z0-9\\s]".r

    val ss = SparkSession
      .builder()
      .appName(name = "task1")
      .config("spark.master", "local[*]")
      .getOrCreate()

    val sc = ss.sparkContext
    sc.setLogLevel("ERROR")

    val inputFile = args(0)

    val rdd = sc.textFile(inputFile)
    val dataRdd =rdd.map(string => parse(string).values.asInstanceOf[Map[String,String]])

    val count = dataRdd.map(x=>x("review_id")).count()

    val year = dataRdd.filter(x =>x("date").contains(args(3).toString)).count()

    val distinct_user = dataRdd.map(x=>x("user_id")).distinct().count()

    val top_reviewers = dataRdd.map(x=>(x("user_id"),1)).reduceByKey((x, y) => x + y).sortBy(_._2,ascending = false).take(args(4).toInt)

    val top_words = dataRdd.flatMap(x=>x("text").toLowerCase.split(" "))
      .filter(x=> ! (pun_list.contains(x)))
      .map(x=>  reg_pattern.replaceAllIn(x,"")).filter(x=> ! (lines.contains(x)))
      .filter(x=> x.length>0).map(x => (x,1)).reduceByKey((x,y)=>x+y).sortBy(x =>(-x._2,x._1)).map(x => x._1).take(args(5).toInt)


    val output_file = new File(args(1))
    val output_writer = new BufferedWriter(new FileWriter(output_file))
    output_writer.write("{"+ "\"A\":"+count.toInt+","+"\"B\":"+year+","+"\"C\":"+distinct_user+","+"\"D\":"+top_reviewers.toList.toString().replaceAll("List","").replaceAll("\\(", "[").replaceAll("\\)", "]")+","+"\"E\":"+top_words.toList.toString().replaceAll("List","").replaceAll("\\(", "[").replaceAll("\\)", "]")+"}")
    output_writer.close()

    sc.stop()

  }

}
